core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointTracker.java

```diff
@@ -173,33 +205,79 @@ synchronized void updateCheckpointOnReplica(final long checkpoint) {
+            }
+        }
+
+        trackingLocalCheckpoints.removeAll(a -> !initializingAllocationIds.contains(a));
+        for (final String a : initializingAllocationIds) {
+            if (inSyncLocalCheckpoints.containsKey(a)) {
+                /*
+                 * This can happen if we mark the allocation ID as in sync at the end of recovery before seeing a cluster state update from
+                 * marking the shard as active.
+                 */
+                continue;
+            }
+            if (trackingLocalCheckpoints.containsKey(a)) {
+                // we are already tracking this allocation ID
+                continue;
             }
+            // this is a new allocation ID
+            trackingLocalCheckpoints.put(a, SequenceNumbersService.UNASSIGNED_SEQ_NO);
+            logger.trace("tracking [{}] via cluster state update from master", a);
         }
-        inSyncLocalCheckpoints.removeAll(key -> assignedAllocationIds.contains(key) == false);
     }
 
     /**
-     * Marks the shard with the provided allocation ID as in-sync with the primary shard. This should be called at the end of recovery where
-     * the primary knows all operations below the global checkpoint have been completed on this shard.
+     * Marks the shard with the provided allocation ID as in-sync with the primary shard. This method will block until the local checkpoint
+     * on the specified shard advances above the current global checkpoint.
      *
-     * @param allocationId the allocation ID of the shard to mark as in-sync
+     * @param allocationId    the allocation ID of the shard to mark as in-sync
+     * @param localCheckpoint the current local checkpoint on the shard
+     *
+     * @throws InterruptedException if the thread is interrupted waiting for the local checkpoint on the shard to advance
      */
-    public synchronized void markAllocationIdAsInSync(final String allocationId) {
-        if (assignedAllocationIds.contains(allocationId) == false) {
-            // master has removed this allocation, ignore
+    public synchronized void markAllocationIdAsInSync(final String allocationId, final long localCheckpoint) throws InterruptedException {
+        if (trackingLocalCheckpoints.containsKey(allocationId) == false) {
             return;
         }
-        logger.trace("marked [{}] as in sync", allocationId);
-        inSyncLocalCheckpoints.put(allocationId, UNASSIGNED_SEQ_NO);
+
+        updateLocalCheckpoint(allocationId, localCheckpoint, trackingLocalCheckpoints, "tracking");
+        waitForAllocationIdToBeInSync(allocationId);
+    }
+
+    private synchronized void waitForAllocationIdToBeInSync(final String allocationId) throws InterruptedException {
+        pendingInSync.add(allocationId);
+        try {
+            while (true) {
+                final long current = trackingLocalCheckpoints.get(allocationId);
```

Comment:
I don't think that we can make any assertions here, we would have to push too much knowledge down and I don't think it's worth it. I can leave a comment explaining the situation though.